cql3, test, utils: switch from boost::adaptors::uniqued to utils::views:unique

tchaikov opened the pull request:
In order to reduce the dependency on external libraries, and for better integration with ranges in C++ standard library. let's use the homebrew `utils::views::unique()` before unique is accepted by the C++ standard.

---

it's a cleanup, hence no need to backport.

regevran commented:
While it is a good initiative, the risk of getting it wrong might shed the need.
Particularity the motivation as presented is not strong enough for me:
> Motivation:
- Boost.Range's uniqued() is incompatible with C++ standard library
  ranges

Where does it meet Scylla? What do we want to do but can't?

> - We need a lightweight solution 

Why do we need it to be lightweight?

> to eliminate all except the first
  element from every consecutive group of equivalent elements.

Why do we need this particular implementation of unique? (why don't we prefer keeping the last element for example and not the first).

tchaikov commented:
> While it is a good initiative, the risk of getting it wrong might shed the need. Particularity the motivation as presented is not strong enough for me:
> 
> > Motivation:
> 
>     * Boost.Range's uniqued() is incompatible with C++ standard library
>       ranges
> 
> 
> Where does it meet Scylla? What do we want to do but can't?

please take a closer at the source code. we have a series of changes which migrate from boost to standard library. i'd prefer not repeating the rationales here.

> 
> > * We need a lightweight solution
> 
> Why do we need it to be lightweight?
> 

for smaller header dependency.

> > to eliminate all except the first
> > element from every consecutive group of equivalent elements.
> 
> Why do we need this particular implementation of unique? (why don't we prefer keeping the last element for example and not the first).

it's explained in the commit message. in short, simpler this way. one would argue why don't we prefer keeping the 2nd element ? or "n / 42" is my favorite element.

nyh commented:
I'm confused about what exists in the C++ standard, and the naming of everything. A std::ranges::unique **does** exist, why can't we use it? Is the reason why you used the name "unique" instead of Boost's "uniqued" because of the C++ standard?

Please mention std::ranges::unique (https://en.cppreference.com/w/cpp/algorithm/ranges/unique) in the commit message and explain why we couldn't use that (if it's a different thing, please explain how it's different).

nyh reviewed (commented):
The implementation looks reasonable although it was too complex for me to be 100% sure I understand it all.

But there is something much more basic I don't understand. The second patch says we need this utils::views::unique **before unique is accepted by the C++ standard**. But, https://en.cppreference.com/w/cpp/algorithm/ranges/unique clames that std::ranges::unique was already accepted in C++20. Is that not true? Or maybe the new "pipe" trick wasn't in C++20 - if it isn't can't you implement this pipe trick only and reuse std::ranges::unique?

tchaikov commented on utils/ranges.hh:
```diff
@@ -0,0 +1,163 @@
+    constexpr auto end() {
+        if constexpr (std::ranges::common_range<V>) {
+            return iterator<false>{std::ranges::end(_base), std::ranges::end(_base)};
+        } else {
```
it's quite popular in STL containers that their `begin()` and `end()` return the same type. but this does not apply to quite a few range views, whose `end()` s return a sentinel, which does not share the same type of `begin()`/

tchaikov commented on utils/ranges.hh:
```diff
@@ -0,0 +1,163 @@
+requires std::ranges::view<V> && std::equality_comparable<std::ranges::range_reference_t<V>>
+class unique_view : public std::ranges::view_interface<unique_view<V>> {
+private:
+    V _base;  // Remove default initialization
```
it's a leftover. i will remove it. i traded for a `requires` in the default constructor in my last revision, but forgot to remove this comment.

tchaikov commented on utils/ranges.hh:
```diff
@@ -0,0 +1,163 @@
+
+template<std::ranges::input_range V>
+requires std::ranges::view<V> && std::equality_comparable<std::ranges::range_reference_t<V>>
+class unique_view : public std::ranges::view_interface<unique_view<V>> {
```
in general, the view adaptors provide two variants

- an adaptor which takes a range view and converts it to another range view. for instance, in https://en.cppreference.com/w/cpp/ranges/transform_view, `std::ranges::transform_view` is an adaptor.
- a closure. which can used as in a pipeline. and `std::views::transform` is a closure

both of which are part of the public interface, but the latter is more popular, i guess.

sure. i will add some doxygen comment.

nyh commented:
> > The implementation looks reasonable although it was too complex for me to be 100% sure I understand it all.
> > But there is something much more basic I don't understand. The second patch says we need this utils::views::unique **before unique is accepted by the C++ standard**. But, https://en.cppreference.com/w/cpp/algorithm/ranges/unique clames that std::ranges::unique was already accepted in C++20. Is that not true?
> 
> `std::ranges::unique()` is a mutator of an existing range. but `boost::adaptors::uniqued` is adaptor which allows one to pipeline range view to it.
> 
> > Or maybe the new "pipe" trick wasn't in C++20
> 
> exactly

Wow, how confusing :-(

So, the comment on "before unique is accepted into the C++ standard" is a bit ambiguous, because "unique" was accepted in one way (std::ranges::unique) but doesn't work in the "pipe" way. Is there any sort of proposal or something asking for a "pipe" version? Do we know what it's API is planned to be, or even that it will be called unique and not uniqued with a d?

tchaikov commented on utils/ranges.hh:
```diff
@@ -0,0 +1,163 @@
+ * SPDX-License-Identifier: LicenseRef-ScyllaDB-Source-Available-1.0
+ */
+
+#pragma once
```
will rename it to `unique_view.hh`. and yes, i plan to add more views in future.

tchaikov commented:
> > > The implementation looks reasonable although it was too complex for me to be 100% sure I understand it all.
> > > But there is something much more basic I don't understand. The second patch says we need this utils::views::unique **before unique is accepted by the C++ standard**. But, https://en.cppreference.com/w/cpp/algorithm/ranges/unique clames that std::ranges::unique was already accepted in C++20. Is that not true?
> > 
> > 
> > `std::ranges::unique()` is a mutator of an existing range. but `boost::adaptors::uniqued` is adaptor which allows one to pipeline range view to it.
> > > Or maybe the new "pipe" trick wasn't in C++20
> > 
> > 
> > exactly
> 
> Wow, how confusing :-(
> 
> So, the comment on "before unique is accepted into the C++ standard" is a bit ambiguous, because "unique" was accepted in one way (std::ranges::unique) but doesn't work in the "pipe" way. Is there any sort of proposal or something asking for a "pipe" version? 

yes, see https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2760r0.html#tier-3

> Do we know what it's API is planned to be, or even that it will be called unique and not uniqued with a d?

The [unique view](https://github.com/ericniebler/range-v3/blob/master/include/range/v3/view/unique.hpp) is currently part of the range-v3 library but has not yet been standardized. According to [P2214](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2760r0.html) "A Plan for C++ Ranges Evolution", it's assigned ["Tier 2"](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2760r0.html#tier-2) priority, which means it's not in the highest priority group for standardization. This suggests it may take some time before we see it in the C++ Standard Library.
If and when it gets accepted into the standard, there's a good chance it will keep the name unique. This naming prediction is supported by the precedent of other range adaptors that were standardized - while Boost used verb participles (like transformed), the C++20 ranges library consistently uses the base verb form (like transform). Following this pattern, unique would be the likely choice over "uniqued".

tchaikov commented:
v2:

- renamed the header to "unique_view.hh"
- renamed the test to "unique_view_test.cc"
- add the doxygen doc for the unique view
- refactored the unique_view's iterator implementation by extracting `find_next()`

nyh reviewed (approved):
Very nice. The new commit message and comments clearly describe what this implements and why it was done this way.

nyh commented:
> #### 🔴 CI State: UNSTABLE

I don't know what this line means so I'm ignoring it. This specific patch passed the CI successfully, I already approve the patch, so I'm merging. If somebody wants to enforce an official code freeze, please send an email to all maintainers. I didn't get such an email, so I assume there is no official code freeze.
